src/eg/src/guardian_share.rs

```diff
@@ -0,0 +1,514 @@
+use std::{borrow::Borrow, iter::zip};
+
+use anyhow::{bail, ensure, Result};
+use num_bigint::BigUint;
+use serde::{Deserialize, Serialize};
+use util::{bitwise::xor, csprng::Csprng, integer_util::to_be_bytes_left_pad};
+
+use crate::{
+    election_parameters::ElectionParameters,
+    guardian::GuardianIndex,
+    guardian_public_key::GuardianPublicKey,
+    guardian_secret_key::GuardianSecretKey,
+    hash::{eg_h, HValue},
+};
+
+/// Encrypted guardian share
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GuardianEncryptedShare {
+    pub dealer: GuardianIndex,
+    pub recipient: GuardianIndex,
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    pub c0: BigUint,
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    pub c1: BigUint,
+    pub c2: HValue,
+}
+
+impl GuardianEncryptedShare {
+    /// This function computes the share encryption secret key as defined in Equation (15)
+    /// The arguments are
+    /// - h_p - the parameter base hash
+    /// - i - the dealer index
+    /// - l - the recipient index
+    /// - capital_k_l - the recipient public key
+    /// - alpha - as in Equation (14)
+    /// - beta - as in Equation (14)
+    fn secret_key(
+        h_p: HValue,
+        i: u32,
+        l: u32,
+        capital_k_l: &BigUint,
+        alpha: &BigUint,
+        beta: &BigUint,
+    ) -> HValue {
+        // v = 0x11 | b(i, 4) | b(l, 4) | b(capital_k, 512) | b(alpha,l, 512) | b(beta,l, 512)
+        let mut v = vec![0x11];
+        v.extend_from_slice(i.to_be_bytes().as_slice());
+        v.extend_from_slice(l.to_be_bytes().as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(capital_k_l, 512).as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(alpha, 512).as_slice());
+        v.extend_from_slice(to_be_bytes_left_pad(beta, 512).as_slice());
+        eg_h(&h_p, &v)
+    }
+
+    /// SHA-256 HMAC this is (currently) the same as eg_h
+    fn hmac(key: &HValue, data: &dyn AsRef<[u8]>) -> HValue {
+        eg_h(key, data)
+    }
```

Comment:
It would be clean to have this as a function in the utils crate without relying on `eg_h`.

Reply:
So why don't you put it there then?